HackAtUCI/irvinehacks-site-2024
# Import Admin site from last year

Importing last year's Admin site to use again this year.

This PR focuses on providing a minimal working version of the Admin site to view applications and submit reviews based on last year's version. Please limit review comments to functional issues, code style (for modifications after the initial import), and relevant questions. We can discuss design and feature improvements separately.

To keep things simple, we are keeping the original client-side data fetching hooks and user identity context for now, but the pages and navigation were updated to use [App Router](https://nextjs.org/docs/app) instead of [Pages Router](https://nextjs.org/docs/pages). While the entire PR is quite lengthy, the first commit is a raw import, and **the remaining changes should be viewed separately**: [Compare abbfd47..a6480f0](https://github.com/HackAtUCI/irvinehacks-site-2024/compare/abbfd471335040288f51dcaaf028de3621e933e1..a6480f0cc3dee3ba79d0cf87a84db7abd618e270).

Resolves #185.

## Changes
- Import last year's Admin site components as-is
  - Import Admin site components of [HackAtUCI/HackUCI-Site](https://github.com/HackAtUCI/HackUCI-Site) exactly as-is
    - Includes page views, subcomponents, hooks, and layout
  - Organize components into approximate locations but without changes
- Install cloudscape-design and swr
  - Install cloudscape-design components and global-styles for Admin site
  - Install swr for client-side data fetching on Admin site
- Fix first-party imports of Admin components
- Fix missing `OptionDefinition` type
  - Extract `MultiselectProps.Option` to use instead of `OptionDefinition`
- Use `UserContext` for Admin site components
  - For now, we are using client-side data fetching, so the user identity cannot be fetched by the server and must be provided through context
- Use dynamic segment params for Applicant UID
  - Replace router with props for dynamic route segment in `Applicant`
- Expose Admin site pages for Applicant(s)
  - Add default exports to `page.tsx` to serve `Applicant`/`Applicants`
  - Include `use client` directive since fetching data from client by hook
- Provide identity through context in Admin layout
  - Add `useUserIdentity` hook for client-side identity on Admin site
  - Provide `UserContext` with identity in layout for Admin site
- Migrate Admin navigation to use App Router hooks
  - Replace old `next/router` hooks with new hooks from `next/navigation`
  - Redesign `followWithNextLink` into hook since `useRouter` is a hook
  - Replace state and effect for `breadcrumbItems` with regular function
- Use Server Component for Admin Layout
  - Metadata cannot be provided from a Client Component
    - Similar reasoning, if we wanted the Admin Layout to be a [Root Layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required)
  - Transfer `UserContext.Provider` to `AdminLayout` which is Client
  - Simplify redirects when not logged in or unauthorized
- Fix `CardHeader` links in Admin `Applicants` view
  - Using `passHref` and `legacyBehavior` does not work with Cloudscape
  - Use Cloudscape `Link` with custom `followWithNextLink` router
  - Simplify event type for `BaseNavigationDetail`
- Update hackathon name in Admin site layout
  - Replace "HackUCI 2023" with "IrvineHacks 2024"
- Update `ApplicationData` fields in Admin site
  - Change `university` to `school` and remove `gender`
  - Update profile links and FRQ question field names
- Serve placeholder page for `AdminDashboard`
  - Remove old `AdminDemo` page
- Reformat Admin site: include all trailing commas
- Fix type error with `fetcher` in `useApplicant`
  - Specify explicit `SWRKey` generic type

## Testing
1. Start the development servers for the site and API
2. Submit some sample applications under various UIDs
3. Add the "reviewer" role to your own user record
4. Acquire a cookie for your own user identity
5. Visit the Admin site at `/admin/dashboard`
6. Navigate to the **Applicants** view from the sidebar
7. Observe the list of applicants is shown
8. Select an applicant by name
9. Observe the application data is shown
